feat: Implement Divided Section Pattern

muhammad-ali-pk opened the pull request:
## Done

- Implemented Divided section pattern
- Added Notifications to Basic section pattern

Fixes [WD-22564](https://warthogs.atlassian.net/browse/WD-22564)

## QA

### Divided Section - New
- Review [Divided section pattern docs](https://vanilla-framework-5623.demos.haus/docs/patterns/divided-section)
- Review [4.31.0 release notes](https://vanilla-framework-5623.demos.haus/docs/whats-new)
- Review [combined examples](https://vanilla-framework-5623.demos.haus/docs/examples/patterns/divided-section/combined) and verify the pattern is structured as per the [Figma](https://www.figma.com/design/1brHCsyeTE6FU7hfJd6ao2/%F0%9F%9A%A7---WIP--Sites---Core-component-library?node-id=673-5104&m=dev)

### Basic Section - Updated
- Review [Divided section pattern's notification block](https://vanilla-framework-5623.demos.haus/docs/patterns/basic-section) at the very bottom, just before Vanilla Macro.



### Check if PR is ready for release

If this PR contains Vanilla SCSS or macro code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [x] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention
  - if existing APIs (CSS classes & macro APIs) are not changed it can be a bugfix release (x.x.**X**)
  - if existing APIs (CSS classes & macro APIs) are changed/added/removed it should be a minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) or macros should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).



[WD-22564]: https://warthogs.atlassian.net/browse/WD-22564?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://vanilla-framework-5623.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/18541)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

muhammad-ali-pk commented:
> It looks like the margin that should apply only to the bullet points is being applied to the entire block. This happens for all bullets (number, bullet, and status).

There isn't any margin on the block. Maybe you meant "padding"?

<img width="244" height="163" alt="image" src="https://github.com/user-attachments/assets/acaf57a0-4e7f-4e1e-8173-5389ac3976e2" />


The padding is also correct according to the design requirements:

<img width="851" height="109" alt="image" src="https://github.com/user-attachments/assets/2a491fb9-13bd-4354-a452-70cad073a127" />

All blocks except the last one (having shallow bottom padding): 

<img width="778" height="315" alt="image" src="https://github.com/user-attachments/assets/dc0676c3-6ab3-4322-a1d8-0064ae0db752" />

The last block (no bottom padding): 

<img width="779" height="435" alt="image" src="https://github.com/user-attachments/assets/9bf20537-6d7b-4664-9204-c2dd25af2b96" />

eliman11 commented:
Thanks both! 

> I wonder if it’s possible to resolve this while keeping the current structure where the list item serves as the whole block.

Sorry just chiming in here @kim-isaac - could we not remove the baseline spacing on the text items to help with alignment? On https://vanillaframework.io/docs/patterns/lists#ordered-lists-with-horizontal-divider for example, the padding is the same way that's implemented here, but the text doesn't have the padding that comes with the h5 styling.

<img width="617" height="420" alt="Screenshot 2025-08-27 at 11 16 08" src="https://github.com/user-attachments/assets/6b36547b-3189-4d69-9ea2-2db98cae7a2f" />

Just one other minor comment - could we change the h5 headings into h3 HTML tags so we don't skip heading hierarchy for accessibility? This will save dev effort in implementing pages so we won't have to ask for this switch in the future!

muhammad-ali-pk commented:
Hello @kim-isaac @eliman11 
I have addressed comments from both of you. Please have a look and let me know your thoughts.

@kim-isaac 
I have added 2 more examples for mixed items, 1 with bulleted headings and 1 with number-bulleted headings.
The difference between the two is that the any list used within the numbered heading variant will be counted relative to the parent heading number, where as for the bulleted heading variant, it will be normal. (It's a feature of Vanilla, when using list items)

Let me show you what I mean:

- List within a bulleted block:


<img width="974" height="780" alt="image" src="https://github.com/user-attachments/assets/8958e853-9945-4b57-a5bd-5cf7d34c7bd6" />


- List within a number-bulleted block:

<img width="974" height="780" alt="image" src="https://github.com/user-attachments/assets/a7684d1b-f9ac-48d6-b92f-2f314dd669d5" />

eliman11 commented:
LGTM, thanks for the quick change @muhammad-ali-pk!

kim-isaac commented:
Thanks for fixing and adding example case @muhammad-ali-pk !! Looks great.

I have one comment left: 
In the bullet and status lists, the rule is indented by 32px instead of spanning the full block. Could we update it to span the full width like the number list?
<img width="655" height="399" alt="image" src="https://github.com/user-attachments/assets/bc1f048a-4824-477a-a7e4-75b45317f02c" />

Thanks again!

muhammad-ali-pk commented:
@kim-isaac Thank you!!

> In the bullet and status lists, the rule is indented by 32px instead of spanning the full block. Could we update it to span the full width like the number list?

Done!

jmuzina commented on templates/docs/examples/patterns/divided-section/top-rule-variants.html:
I think the "default top rule" title text should change from example to example here

Default -> highlighted -> muted -> no

<img width="1396" height="659" alt="Image" src="https://github.com/user-attachments/assets/d8e468db-36f2-4638-a5cf-9e2d988da09a" />

jmuzina commented on templates/docs/patterns/divided-section/index.md:
```diff
@@ -0,0 +1,344 @@
+A divided section is used to create structured content section with a title, a description block and various divided blocks. It
+provides a flexible layout system that can accommodate different types of content including text, images, videos, lists,
+code blocks, logos, and call-to-action elements, within both the description and divided blocks.
+Internally, it utilizes <a href="/docs/patterns/basic-section">Basic section pattern</a>, so the configuration for most of the layout options, as well as for content items is going to be identical.
```
This is really interesting. I see you say that this pattern is using the basic section pattern internally, but it's not really calling the basic section pattern - it's calling the basic section item block macro to allow you to place the same blocks in the divided section pattern.

For later, out of scope of this PR: I wonder if the basic section block/items should be pulled out of the basic section pattern and considered to be smaller patterns of their own. It would make more sense and be more maintainable if patterns didn't use internal blocks of other patterns. This could be a cleanup task that could also allow those block to be used outside of larger, section-scoped patterns. CC @kim-isaac

muhammad-ali-pk replied:
> I wonder if the basic section block/items should be pulled out of the basic section pattern and considered to be smaller patterns of their own. 

I agree!

> This is really interesting. I see you say that this pattern is using the basic section pattern internally, but it's not really calling the basic section pattern - it's calling the basic section item block macro (which was formerly an internal implementation detail of the basic section only) to allow you to place the same blocks in the divided section pattern.

For now, would you suggest changing the wording?

britneywwc commented on templates/_macros/vf_basic-section.jinja:
```diff
@@ -255,9 +272,52 @@
+    {%- if top_rule_variant != "default" -%}
+      {#-
+        p-rule--highlighted doesn't exist (use p-rule--highlight instead), but p-rule--muted does.
+        We keep the external API here consistent ("-ed" suffix) for simplicity but need to handle this internally.
```
Are there plans to standardise the naming convention for variants like these? If not, would it make sense to rename the variant as `highlight` for consistency?

muhammad-ali-pk replied:
I believe we handle this internally to make sure the external API for top rule variant is consistent (i.e., muted, highlighted)
Though I agree we should standardize so either both classes are without "ed" e.g., `p-rule--mute`, `p-rule--highlight` or both are with "ed" e.g., `p-rule--muted`, `p-rule--highlighted`
(cc: @jmuzina )

britneywwc replied:
Thank you Julie!

britneywwc reviewed (commented):
Thank you, left some comments/questions! 

One thing to note is that once this PR is merged, the new pages should be added to Google Search Console to allow the search module to index these pages when queried

muhammad-ali-pk commented on templates/docs/examples/patterns/divided-section/bulleted-list.html:
```diff
@@ -0,0 +1,73 @@
+              {
+                "type": "cta-block",
+                "item": {
+                  "secondaries": [
```
The CTA block can have different combination of primary, secondary and link texts.
I believe primary button's presence is not mandatory, and that since there can be more than one secondary buttons, @jmuzina chose to name it as "secondaries" which makes sense to me.
(cc: @jmuzina)

muhammad-ali-pk commented:
@jmuzina Can you please have a approve the percy build and merge this PR, if there are no comments from your side?
